### ISSUE TITLE
Multiple `SetValueCurveAtTime` overlapping

### ISSUE THREAD

**Issue by padenot** [WebAudio/web-audio-api#1721]
> If setValueCurveAtTime() is called for time T and duration D and there are any events having a time greater than T, but less than T+D, then a NotSupportedError exception MUST be thrown. In other words, it’s not ok to schedule a value curve during a time period containing other events.
> 
> Similarly a NotSupportedError exception MUST be thrown if any automation method is called at a time which is inside of the time interval of a SetValueCurve event at time T
> and duration D

```
var ac = new AudioContext()
var g = ac.createGain()
g.gain.setValueCurveAtTime([1,2], 4, 4); 
g.gain.setValueCurveAtTime([1,4], 4, 2); 
```

Is the last call supposed to throw or not? i.e., in the first sentence, is it "having a time greater than T", "having a time greater or equal than T", "having a time strictly greater than T" ?

In the second sentence, I find "inside" to not be precise enough. In other words, is it _[T,T+D]_ or _(T, T+D)_.

I'm thinking that it should throw, because I don't like those overlapping curves.

**Comment by rtoy** [comment]
I agree.  But I think it needs a bit of refinement.  If we have
```
g.gain.setValueCurveAtTime([1,2],4,4);
```
it's ok to then do `g.gain.setValueAtTime(42, 4+4)`.  That is, you can schedule a new event at the end of the curve.  So the condition is probably [T, T+D).

(In fact, this is how Chrome implements setValueCurve: Chrome automatically appends setValueAtTime(interpolated-curve-end-value, T+D).  Because we had a bug where the value after the curve wasn't the final curve value.

**Comment by padenot** [comment]
Right, I'm in the process of implementing it properly in Firefox as well, and I did exactly the same, it's allowed to put an event exactly at the end of the curve.

I'm proposing to rephrase the section like so:

> If setValueCurveAtTime() is called for time T and duration D and there are any events having a time 
> greater or equal to T, but less than T+D, then a NotSupportedError exception MUST be thrown. In 
> other words, it’s not ok to schedule a value curve with the start time of another event, or during a 
> time period containing other events. It's ok to schedule a curve that ends at another event.
> 
>  Similarly a NotSupportedError exception MUST be thrown if any automation method is called at a 
> time which is inside [T, D). In other word, it's not ok to add an event at the start of in the middle of a 
> curve, but it's allowed to add an event at the very end of the curve.

Thoughts ?

**Comment by rtoy** [comment]
Looks good with a nit: "inside [T,D)" is probably better phrased as "is contained in the interval [T,D)".

**Comment by karlt** [comment]
The second paragraph looks good to me, thanks (with `s/D)/T+D)/`,
`s/word/words/` and `s/of in/of or in/`).

> In other words, it’s not ok to schedule a value curve with the start time of
> another event

I don't think we need to exclude scheduling a value curve with a start time
matching the time of another event.  This would desireable after a
`linearRampToValueAtTime` at least, and could and IMO should be acceptable for
all other events except `setValueCurveAtTime`.  This would match the situation
for scheduling other event types, which may be scheduled at the same time as
any another event except `setValueCurveAtTime`.

Are there any cases that are not sufficiently specified with the modification
to the second paragraph?

Would it help to clarify the first paragraph with "A value curve `startTime`
may match the time of any other event except another value curve"?

**Comment by rtoy** [comment]
I don't follow how that would work, assuming you meant that this would be allowed:

```
setValueAtTime(0, 0);
linearRampToValueAtTime(99, 1);
setValueCurveAtTime([1, 2], 1, 1);
```
What is the value at time t = 1?  The linear ramp formula says it should be 99, but the setValueCurve says it should be 1.

I guess we have a similar issue if you did
```
linearRampToValueAtTime(99, 1);
setValueAtTime(0, 1);
```
What is the value of the audio param at time 1? 99? 0?

**Comment by karlt** [comment]
The events are ordered according to "If one of these events is added at a time
where there is already one or more events, then it will be placed in the list
after them".

At time 1, each event would have been processed in order.  The most recent
would define the value at time 1 (if it is not a setTarget event).

With `setValueCurveAtTime([1, 2], 1, 1)`, the value would be 1.
With `setValueAtTime(0, 1)`, the value would be 0.

This is consistent with the behavior for two setValueAtTime events at the same
time, expected by
https://github.com/web-platform-tests/wpt/commit/575c6d120e84e2dd1343bd42c95f7d6b2477c12a#diff-8dc0400d43692e2c4fe53a5cd3b6c625R31
The reasoning may be a little different from what was imagined when that
comment was written, but I think we can still expect mostly the same values
at the event times.

I assume we are all in agreement that the second (overlapping) setValueCurve in
that test should now throw.

**Comment by padenot** [comment]
This makes sense and is in line with everything else.

**Comment by padenot** [comment]
In fact, I'm having a look at [existing WPTs](https://github.com/web-platform-tests/wpt/blob/master/webaudio/the-audio-api/the-audioparam-interface/audioparam-setValueCurve-exceptions.html#L220-L239), and seem to be allowed already.

**Comment by rtoy** [comment]
Agreed.

**Comment by padenot** [comment]
Closed by #1723.
